include/RAJA/pattern/params/forall.hpp

```diff
@@ -0,0 +1,320 @@
+    }
+    template<typename EXEC_POL, typename... Params, typename ...Args, typename FP = ForallParamPack<Params...>>
+    static void constexpr resolve( ForallParamPack<Params...>& f_params, Args&& ...args){
+      FP::detail_resolve(EXEC_POL(), typename FP::params_seq(), f_params, std::forward<Args>(args)... );
+    }
+  };
+  //===========================================================================
+
+
+
+  //===========================================================================
+  //
+  //
+  // ForallParamPack generators.
+  //
+  //
+  RAJA_INLINE static auto get_empty_forall_param_pack(){
+    static ForallParamPack<> p;
+    return p;
+  }
+
+  namespace detail {
+    // all_true trick to perform variadic expansion in static asserts.
+    // https://stackoverflow.com/questions/36933176/how-do-you-static-assert-the-values-in-a-parameter-pack-of-a-variadic-template
+    template<bool...> struct bool_pack;
+    template<bool... bs>
+    using all_true = std::is_same<bool_pack<bs..., true>, bool_pack<true, bs...>>;
+
+    template<typename Base, typename... Ts>
+    using check_types_derive_base = all_true<std::is_convertible<Ts, Base>::value...>;
+  } // namespace detail
+
+
+  template<typename... Ts>
+  constexpr auto make_forall_param_pack_from_tuple(const camp::tuple<Ts...>& tuple) {
+    static_assert(detail::check_types_derive_base<detail::ForallParamBase, Ts...>::value,
+        "Forall optional arguments do not derive ForallParamBase. Please see Reducer, ReducerLoc and KernelName for examples.") ;
+    return ForallParamPack<Ts...>(tuple);
+  }
+
+  
+
+  namespace detail {
+    // Maybe we should do a lot of these with structs...
+    template<camp::idx_t... Seq, typename TupleType>
+    constexpr auto tuple_from_seq (const camp::idx_seq<Seq...>&, const TupleType& tuple){
+      return camp::make_tuple( camp::get< Seq >(tuple)... );
+    };
+
+    template<typename... Ts>
+    constexpr auto strip_last_elem(camp::tuple<Ts...>&& tuple){
+      return tuple_from_seq(camp::make_idx_seq_t<sizeof...(Ts)-1>{},tuple);
+    };
+  } // namespace detail
+
+
+  // Make a tuple of the param pack except the final element...
+  template<typename... Args>
+  constexpr auto make_forall_param_pack(Args&&... args){
+    // We assume the last element of the pack is the lambda so we need to strip it from the list.
+    auto stripped_arg_tuple = detail::strip_last_elem(camp::make_tuple(args...)); 
+    return make_forall_param_pack_from_tuple(stripped_arg_tuple);
+  }
+  //===========================================================================
+
+
+
+  //===========================================================================
+  //
+  //
+  // Callable should be the last argument in the param pack, just extract it...
+  //
+  //
+  template<typename... Args>
+  constexpr auto get_lambda(Args&&... args){
+    auto lambda = camp::get<sizeof...(Args)-1>( camp::make_tuple(args...) );
+    return lambda; 
```

Comment:
Comdining my two previous comments as a suggestion.
```suggestion
  constexpr decltype(auto) get_lambda(Args&&... args){
    return camp::get<sizeof...(Args)-1>( camp::forward_as_tuple(std::forward<Args>(args)...) );
```